src/lib.rs

```diff
@@ -633,6 +634,373 @@ pub extern "C" fn delete(
+    fn success(self, entries: Vec<ObjectMeta>) { // TODO entries
+        let entries_slice = entries.into_iter()
+            .map(Into::into)
+            .collect::<Vec<_>>()
+            .into_boxed_slice();
+
+        let entry_count = entries_slice.len() as u64;
+        let entries_ptr = entries_slice.as_ptr();
+        std::mem::forget(entries_slice);
+
+        self.response.result = CResult::Ok;
+        self.response.entry_count = entry_count;
+        self.response.entries = entries_ptr;
+        self.response.error_message = std::ptr::null_mut();
+    }
+}
+
+impl Drop for ListResponseGuard {
+    fn drop(&mut self) {
+        self.notify_on_drop()
+    }
+}
+
+unsafe impl Send for ListResponseGuard {}
+
+#[no_mangle]
+pub extern "C" fn destroy_list_entries(
+    entries: *mut ListEntry,
+    entry_count: u64
+) -> CResult {
+    let boxed_slice = unsafe { Box::from_raw(std::slice::from_raw_parts_mut(entries, entry_count as usize)) };
+    for entry in &*boxed_slice {
+        // Safety: must properly drop all allocated fields from ListEntry here
+        let _ = unsafe { CString::from_raw(entry.location.cast_mut()) };
+        if !entry.e_tag.is_null() {
+            let _ = unsafe { CString::from_raw(entry.e_tag.cast_mut()) };
+        }
+        if !entry.version.is_null() {
+            let _ = unsafe { CString::from_raw(entry.version.cast_mut()) };
+        }
+    }
+    CResult::Ok
+}
+
+impl From<object_store::ObjectMeta> for ListEntry {
+    fn from(meta: object_store::ObjectMeta) -> Self {
+        ListEntry {
+            location: CString::new(meta.location.to_string())
+                .expect("should not have nulls")
+                .into_raw(),
+            last_modified: meta.last_modified
+                .timestamp()
+                .try_into()
+                .expect("is positive"),
+            size: meta.size as u64,
+            e_tag: match meta.e_tag {
+                None => std::ptr::null(),
+                Some(s) => {
+                    CString::new(s)
+                        .expect("should not have nulls")
+                        .into_raw()
+                }
+            },
+            version: match meta.version {
+                None => std::ptr::null(),
+                Some(s) => {
+                    CString::new(s)
+                        .expect("should not have nulls")
+                        .into_raw()
+                }
+            }
+        }
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn list(
+    prefix: *const c_char,
+    config: *const Config,
+    response: *mut ListResponse,
+    handle: *const c_void
+) -> CResult {
+    let response = unsafe { ListResponseGuard::new(response, handle) };
+    let prefix = unsafe { std::ffi::CStr::from_ptr(prefix) };
+    let prefix: Path = prefix.to_str().expect("invalid utf8").try_into().unwrap();
+    let config = unsafe { & (*config) };
+
+    match RT.get() {
+        Some(runtime) => {
+            runtime.spawn(async move {
```

Comment:
Shouldn't we use our `buffer_unordered` dispatch queue?

Reply:
I think it is better to leave the low throughput operations out of the queue in order to not introduce latency due to a bunch of get or put requests that may be happening in the background. Unless we are expecting hundreds of list operations per second, then it should be going through the queue, what do you think?